### PR TITLE
Add Lightning Data README 1/N

### PR DESCRIPTION
**Pull request by tchaton**
## What does this PR do?

This PR adds the first iteration of Lightning Data README.

Fixes #\<issue_number>

<details>
  <summary><b>Before submitting</b></summary>

- Was this **discussed/agreed** via a GitHub issue? (not for typos and docs)
- [ ] Did you read the [contributor guideline](https://github.com/Lightning-AI/lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [ ] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- Did you make sure to **update the documentation** with your changes? (if necessary)
- Did you write any **new necessary tests**? (not for typos and docs)
- [ ] Did you verify new and **existing tests pass** locally with your changes?
- Did you list all the **breaking changes** introduced by this pull request?
- Did you **update the CHANGELOG**? (not for typos, docs, test updates, or minor internal changes/refactors)

</details>

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing, make sure you have read the [review guidelines](https://github.com/Lightning-AI/lightning/wiki/Review-guidelines). In short, see the following bullet-list:

<details>
  <summary>Reviewer checklist</summary>

- [ ] Is this pull request ready for review? (if not, please submit in draft mode)
- [ ] Check that all items from **Before submitting** are resolved
- [ ] Make sure the title is self-explanatory and the description concisely explains the PR
- [ ] Add labels and milestones (and optionally projects) to the PR so it can be classified

</details>

----
📚 Documentation preview 📚: https://pytorch-lightning--19455.org.readthedocs.build/en/19455/

cc @borda

### CODE REVIEW

**Review by awaelchli — APPROVED** [review]
I believe the fault-tolerant aspect could be highlighted as well, which is usually a huge pain point.

File: src/lightning/data/README.md [hunk 1]
```diff
@@ -0,0 +1,359 @@
...
+for i in range(1000):
+    width = np.random.randint(224, 320) 
+    height = np.random.randint(224, 320) 
+    image_path = os.path.join(data_dir, f"{i}.JPEG")
+    Image.fromarray(
+        np.random.randint(0, 256, (width, height, 3), np.uint8)
+    ).save(image_path, format="JPEG", quality=90)
+```
+
+```python
```
Comment (awaelchli) [inline src/lightning/data/README.md:128]:
```suggestion
```bash
```

File: src/lightning/data/README.md [hunk 2]
```diff
@@ -0,0 +1,359 @@
...
+
+```
+
+### 2. Upload Your Data to Cloud Storage
+
+Cloud providers such as [AWS](https://docs.aws.amazon.com/cli/latest/userguide/getting-started-install.html), [Google Cloud](https://cloud.google.com/storage/docs/uploading-objects?hl=en#upload-object-cli), [Azure](https://learn.microsoft.com/en-us/azure/import-export/storage-import-export-data-to-files?tabs=azure-portal-preview), etc.. provide command line client to upload your data to their storage.
+
+Here is an example with [AWS S3](https://aws.amazon.com/s3).
+
+```python
```
Comment (awaelchli) [inline src/lightning/data/README.md:78]:
```suggestion
```bash
```

File: src/lightning/data/README.md [hunk 3]
```diff
@@ -0,0 +1,371 @@
...
+| [SlimPajama](https://huggingface.co/datasets/cerebras/SlimPajama-627B) & [StartCoder](https://huggingface.co/datasets/bigcode/starcoderdata) |        Text         |              [Prepare the TinyLlama 1T token dataset](https://lightning.ai/lightning-ai/studios/prepare-the-tinyllama-1t-token-dataset) |
+| Generated                                                                                                                                    |    Parquet Files    |            [Convert parquets to Lightning Streaming](https://lightning.ai/lightning-ai/studios/convert-parquets-to-lightning-streaming) |
+
+[Lightning Studios](https://lightning.ai) are fully reproducible cloud IDE with data, code, dependencies, etc... Finally reproducible science.
+
+# 📈 Easy Data Processing Scaling
+
+The easiest way to scale is to create a free account on [lightning.ai](https://lightning.ai/) platform. Using the platform, the `optimize` and `map` can start multiple machines to make data processing drastically faster.
```
Comment (nohalon) [inline src/lightning/data/README.md]: I dont think this sentence is correct

> Using the platform, the `optimize` and `map` can start multiple machines to make data processing drastically faster.

File: src/lightning/data/README.md [hunk 4]
```diff
@@ -0,0 +1,371 @@
...
+
+The Data Prep Job UI from the [LAION 400M Studio](https://lightning.ai/lightning-ai/studios/use-or-explore-laion-400million-dataset) where we used 32 machines to download 400 million images in only 2 hours.
+
+</div>
+
+# 🔑 Key Features
+
+## Multi-GPU / Multi-Node
+
+You have nothing to do, the StreamingDataset takes care of everything for you. It automatically make sure each rank receives different batch of data.
```
Comment (nohalon) [inline src/lightning/data/README.md:210]:
```suggestion
The StreamingDataset takes care of everything for you. It automatically make sure each rank receives different batch of data.
```

File: src/lightning/data/README.md [hunk 5]
```diff
@@ -0,0 +1,371 @@
...
+input_dir = "tinyllama-data"
+train_datasets = [
+    StreamingDataset(
+        input_dir="tinyllama-data/slimpajama/train",
+        item_loader=TokensLoader(block_size=effective_block_size),
+        shuffle=True,
+        drop_last=True,
+    ),
+    StreamingDataset(
+        input_dir="tinyllama-data/starcoder/",
```
Comment (nohalon) [inline src/lightning/data/README.md:267]:
```suggestion
        input_dir="tinyllama-data/starcoder",
```
either remove the slash here or add it above for consistency

File: src/lightning/data/README.md [hunk 6]
```diff
@@ -0,0 +1,371 @@
...
+# Iterate over the combined datasets
+for batch in tqdm(train_dataloader):
+    pass
+```
+
+## Stateful StreamingDataLoader
+
+Lightning Data provides a stateful `StreamingDataLoader`. This simplifies resuming training over large datasets.
+
+Note: This is used by [Lit-GPT](https://github.com/Lightning-AI/lit-gpt/blob/main/pretrain/tinyllama.py) to pretrain LLMs.
```
Comment (nohalon) [inline src/lightning/data/README.md:288]:
```suggestion
Note: `StreamingDataLoader` is used by [Lit-GPT](https://github.com/Lightning-AI/lit-gpt/blob/main/pretrain/tinyllama.py) to pretrain LLMs.
```

File: src/lightning/data/README.md [hunk 7]
```diff
@@ -0,0 +1,371 @@
...
+for batch_idx, batch in enumerate(dataloader):
+  
+    # Store the state every 1000 batches
+    if batch_idx % 1000 == 0:
+        torch.save(dataloader.state_dict(), "dataloader_state.pt")
+```
+
+## Profiling
+
+The `StreamingDataLoader` supports profiling your dataloading. Simply pass the following `profile_batches` argument as follows:
```
Comment (nohalon) [inline src/lightning/data/README.md:316]:
```suggestion
The `StreamingDataLoader` supports profiling your data loading. Simply use the `profile_batches` argument as follows:
```

File: src/lightning/data/README.md [hunk 8]
```diff
@@ -0,0 +1,371 @@
...
+
+```python
+from lightning.data import StreamingDataset
+
+dataset = StreamingDataset(..., max_cache_size="10GB")
+```
+
+# ⚡ Contributors
+
+We welcome any contributions, pull requests, or issues. If you use the Streaming Dataset for your own project, please reach out to us on Slack.
```
Comment (nohalon) [inline src/lightning/data/README.md:372]: Discord instead of Slack?

File: src/lightning/data/README.md [hunk 9]
```diff
@@ -0,0 +1,371 @@
...
+<br/>
+<br/>
+
+## Blazing fast, distributed streaming of training data from cloud storage
+
+</div>
+
+# ⚡ Welcome to Lightning Data
+
+We developed `Streaming Dataset` to optimize training of large datasets from cloud storage, prioritizing speed, affordability, and scalability.
```
Comment (nohalon) [inline src/lightning/data/README.md:14]:
```suggestion
We developed `StreamingDataset` to optimize training of large datasets stored on the cloud while prioritizing speed, affordability, and scalability.
```

File: src/lightning/data/README.md [hunk 10]
```diff
@@ -0,0 +1,371 @@
...
+
+Specifically crafted for multi-node, distributed training with large models, it enhances accuracy, performance, and user-friendliness. Now, training efficiently is possible regardless of the data's location. Simply stream in the required data when needed.
+
+The `Streaming Dataset` is compatible with any data type, including **images, text, video, and multimodal data** and it is a drop-in replacement for your PyTorch [IterableDataset](https://pytorch.org/docs/stable/data.html#torch.utils.data.IterableDataset) class. For example, it is used by [Lit-GPT](https://github.com/Lightning-AI/lit-gpt/blob/main/pretrain/tinyllama.py) to pretrain LLMs.
```
Comment (nohalon) [inline src/lightning/data/README.md:18]:
```suggestion
The `StreamingDataset` is compatible with any data type, including **images, text, video, and multimodal data** and it is a drop-in replacement for your PyTorch [IterableDataset](https://pytorch.org/docs/stable/data.html#torch.utils.data.IterableDataset) class. For example, it is used by [Lit-GPT](https://github.com/Lightning-AI/lit-gpt/blob/main/pretrain/tinyllama.py) to pretrain LLMs.
```

File: src/lightning/data/README.md [hunk 11]
```diff
@@ -0,0 +1,371 @@
...
+
+Finally, the `Streaming Dataset` is fast, check out our [benchmark](https://lightning.ai/lightning-ai/studios/benchmark-cloud-data-loading-libraries).
```
Comment (nohalon) [inline src/lightning/data/README.md:20]:
```suggestion
Finally, the `StreamingDataset` is fast! Check out our [benchmark](https://lightning.ai/lightning-ai/studios/benchmark-cloud-data-loading-libraries).
```

File: src/lightning/data/README.md [hunk 12]
```diff
@@ -0,0 +1,371 @@
...
+
+Here is an illustration showing how the Streaming Dataset works.
```
Comment (nohalon) [inline src/lightning/data/README.md:22]:
```suggestion
Here is an illustration showing how the `StreamingDataset` works.
```

File: src/lightning/data/README.md [hunk 13]
```diff
@@ -0,0 +1,371 @@
...
+
+```bash
+pip install --no-cache-dir git+https://github.com/Lightning-AI/pytorch-lightning.git@master
+```
+
+## Quick Start
+
+### 1. Prepare Your Data
+
+You can convert your raw dataset into Lightning Streaming format using the `optimize` operator. More formats are coming...
```
Comment (nohalon) [inline src/lightning/data/README.md:42]:
```suggestion
Convert your raw dataset into Lightning Streaming format using the `optimize` operator. More formats are coming...
```

**Review by nohalon — APPROVED** [review]
some comments

File: src/lightning/data/README.md (hunk 6 above)
Comment (tchaton) [reply src/lightning/data/README.md:256]: done.

File: src/lightning/data/README.md (hunk 8 above)
Comment (tchaton) [reply src/lightning/data/README.md:372]: done.

File: src/lightning/data/README.md (hunk 3 above)
Comment (tchaton) [reply src/lightning/data/README.md]: done.
